test/__snapshots__/test.js.snap

```diff
@@ -1,5 +1,149 @@
+const utils = require(\\"./utils.js\\");
+
+exports.convert = (value, { context = \\"The provided value\\" } = {}) => {
+  if (typeof value !== \\"function\\") {
+    throw new TypeError(context + \\" is not a function\\");
+  }
+
+  function invokeTheCallbackFunction(...args) {
+    if (new.target !== undefined) {
+      throw new Error(\\"Internal error: invokeTheCallbackFunction is not a constructor\\");
+    }
+
+    const thisArg = utils.tryWrapperForImpl(this);
+    let callResult;
+
+    for (let i = 0; i < args.length; i++) {
+      args[i] = utils.tryWrapperForImpl(args[i]);
+    }
+
+    callResult = Reflect.apply(value, thisArg, args);
+
+    callResult = conversions[\\"any\\"](callResult, { context: context });
+
+    return callResult;
+  }
+
+  invokeTheCallbackFunction.construct = (...args) => {
+    for (let i = 0; i < args.length; i++) {
+      args[i] = utils.tryWrapperForImpl(args[i]);
+    }
+
+    let callResult = Reflect.construct(value, args);
+
+    callResult = conversions[\\"any\\"](callResult, { context: context });
+
+    return callResult;
+  };
+
+  invokeTheCallbackFunction[utils.wrapperSymbol] = value;
+  invokeTheCallbackFunction.objectReference = value;
+
+  return invokeTheCallbackFunction;
+};
+"
+`;
+
+exports[`built-in types VoidFunction 1`] = `
+"\\"use strict\\";
+
+const conversions = require(\\"webidl-conversions\\");
+const utils = require(\\"./utils.js\\");
+
+exports.convert = (value, { context = \\"The provided value\\" } = {}) => {
+  if (typeof value !== \\"function\\") {
+    throw new TypeError(context + \\" is not a function\\");
+  }
+
+  function invokeTheCallbackFunction() {
+    if (new.target !== undefined) {
+      throw new Error(\\"Internal error: invokeTheCallbackFunction is not a constructor\\");
+    }
+
+    const thisArg = utils.tryWrapperForImpl(this);
+    let callResult;
+
+    callResult = Reflect.apply(value, thisArg, []);
+  }
+
+  invokeTheCallbackFunction.construct = () => {
+    let callResult = Reflect.construct(value, []);
+  };
+
+  invokeTheCallbackFunction[utils.wrapperSymbol] = value;
+  invokeTheCallbackFunction.objectReference = value;
+
+  return invokeTheCallbackFunction;
+};
+"
+`;
+
+exports[`with processors AsyncCallbackFunction.webidl 1`] = `
+"\\"use strict\\";
+
+const conversions = require(\\"webidl-conversions\\");
+const utils = require(\\"./utils.js\\");
+
+exports.convert = (value, { context = \\"The provided value\\" } = {}) => {
+  if (typeof value !== \\"function\\") {
+    throw new TypeError(context + \\" is not a function\\");
+  }
+
+  function invokeTheCallbackFunction() {
+    if (new.target !== undefined) {
+      throw new Error(\\"Internal error: invokeTheCallbackFunction is not a constructor\\");
+    }
+
+    const thisArg = utils.tryWrapperForImpl(this);
+    let callResult;
+
+    try {
+      callResult = Reflect.apply(value, thisArg, []);
+
+      callResult = Promise.resolve(callResult).then(
```

Comment:
This doesn't implement the semantics described in https://heycam.github.io/webidl/#es-invoking-callback-functions + https://heycam.github.io/webidl/#es-promise . For that the correct code would be something more like

```js
return new Promise(resolve => resolve(callResult))
```

Note that `Promise.resolve()` is inappropriate because it returns its argument in some cases, which is never true for converting to a Web IDL promise type. And note how converting to a Web IDL `Promise<T>` type ignores the `T`.

Reply:
This code is generated by: <https://github.com/jsdom/webidl2js/blob/7d45dfc6e6d1ba737b77eaabbab712dcf2d19435/lib/types.js#L330-L349>

Changing that is outside the scope of this PR.

Reply:
Well, I'm not willing to merge this PR if it generates incorrect code that is unusable for consumers of webidl2js, so I guess fixing that is a dependency of this PR then.

Reply:
Done in <https://github.com/jsdom/webidl2js/pull/219>.